drivers/bluetooth/hci/h4.c

```diff
@@ -159,15 +161,82 @@ static void reset_rx(void)
 
 static struct net_buf *get_rx(int timeout)
 {
+	struct net_buf *buf;
+
 	BT_DBG("type 0x%02x, evt 0x%02x", rx.type, rx.evt.evt);
 
 	if (rx.type == H4_EVT) {
-		return bt_buf_get_evt(rx.evt.evt, timeout);
+		buf = bt_buf_get_evt(rx.evt.evt, timeout);
+	} else {
+		buf = bt_buf_get_rx(BT_BUF_ACL_IN, timeout);
+	}
+
+	if (buf) {
+		if (rx.discardable) {
+			buf->flags |= BUF_DISCARDABLE;
+		} else {
+			buf->flags &= ~BUF_DISCARDABLE;
+		}
 	}
 
-	return bt_buf_get_rx(BT_BUF_ACL_IN, timeout);
+	return buf;
 }
 
+#if (CONFIG_BT_H4_DISCARD_RX_WAIT >= 0)
+static struct net_buf *get_rx_blocking(void)
+{
+	sys_slist_t list = SYS_SLIST_STATIC_INIT(&list);
+	bool discarded = false;
+	struct net_buf *buf;
+
+	/* If we have ACL flow control enabled then ACL buffers come from a
+	 * dedicated pool, and we cannot benefit from trying to discard
+	 * events from the RX queue.
+	 */
+	if (IS_ENABLED(CONFIG_BT_HCI_ACL_FLOW_CONTROL) && rx.type == H4_ACL) {
+		return get_rx(K_FOREVER);
+	}
+
+	buf = get_rx(CONFIG_BT_H4_DISCARD_RX_WAIT);
+	if (buf) {
+		return buf;
+	}
+
+	BT_WARN("Attempting to discard packets from RX queue");
+
+	while ((buf = net_buf_get(&rx.fifo, K_NO_WAIT))) {
+		if (!discarded && (buf->flags & BUF_DISCARDABLE)) {
```

Comment:
So this seems to discard only one buffer at time? Is that correct behavior perhaps you can check if if there have been anything added to the list, if there isn't anything that means the order is preserved and we can return interrupting the while loop since you don't have to call k_fifo_put_slist.

Reply:
Not quite. Even if we don't find anything to discard we'll have emptied rx.fifo through the net_buf_get calls. So we anyway have to do the put_slist(). And this code is discarding at most one buffer to minimise packet loss for something like mesh that's waiting for advertising reports.